Add support for Blob

joanlopez opened the pull request:
## What?

It adds support for [Blob](https://developer.mozilla.org/en-US/docs/Web/API/Blob), as described in #35 

## Why?

In order to stabilize the API before considering the extension non-experimental.

## Checklist

<!-- 
If you haven't read the contributing guidelines https://github.com/grafana/k6/blob/master/CONTRIBUTING.md 
and code of conduct https://github.com/grafana/k6/blob/master/CODE_OF_CONDUCT.md yet, please do so
-->

- [ ] I have performed a self-review of my code.
- [ ] I have added tests for my changes.
- [ ] I have run linter locally (`make lint`) and all checks pass.
- [ ] I have run tests locally (`make test`) and all tests pass.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
<!-- - [ ] Any other relevant item -->

## Related PR(s)/Issue(s)

<!-- - <https://github.com/grafana/...> -->

<!-- Does it close an issue? -->

Closes #35 

<!-- Thanks for your contribution! 🙏🏼 -->

joanlopez commented on websockets/blob.go:
```diff
@@ -0,0 +1,75 @@
+	"go.k6.io/k6/js/common"
+)
+
+type Blob struct {
```
I'm not 100% if we do really want this to remain specific to the extension, or to be available in k6 as another experimental package (similar to `streams`).
cc/ @mstoykov 

To be honest, I developed this within k6 codebase in order to reuse all the WPTs set up, at least while coding. So, I'm fine with any approach. If we prefer to keep it here, I can add a few tests specifically for `Blob` manually.

mstoykov replied:
As I said internally - IMO moving this to core *now* will slow down the stabilization of the extension.

Which happens to be the reason we are implementing this to begin with.

I think it will be a lot easier to try to move this to core and usable in other places *after* the extension is out of experimental.

joanlopez replied:
Great then, we can move forward as-is - thanks! 🙇🏻

joanlopez commented on websockets/blob.go:
```diff
@@ -0,0 +1,75 @@
+	return b.data.String()
+}
+
+func (r *WebSocketsAPI) blob(call sobek.ConstructorCall) *sobek.Object {
+	var blobParts []interface{}
+	if len(call.Arguments) > 0 {
+		if parts, ok := call.Arguments[0].Export().([]interface{}); ok {
+			blobParts = parts
+		}
+	}
+
+	return newBlob(r.vu.Runtime(), blobParts)
+}
```
That's also another open question: do we really want the `Blob` constructor to be available widely? Otherwise I guess we can just remove this.

mstoykov replied:
I don't see a reason why not to and given that people will likely also want to be able to use at as a thing to *send* over the websocket it seems like a requirement.

joanlopez replied:
Super! 🎉

joanlopez commented on websockets/websockets.go:
```diff
@@ -425,9 +424,6 @@ func (w *webSocket) loop() {
 	}
 }
 
-const binarytypeWarning = `You have not set a Websocket binaryType to "arraybuffer", but you got a binary response. ` +
```
As the warning was already there, and we're moving towards making the extension non-experimental, and spec compliant, I guess it's fine to just add this "breaking change" (changing the default behavior), but I'm open for suggestions.

mstoykov commented on websockets/websockets.go:
```diff
@@ -448,13 +444,15 @@ func (w *webSocket) queueMessage(msg *message) {
-			if w.binaryType == "" {
-				w.binaryType = arraybufferBinaryType
-				w.vu.State().Logger.Warn(binarytypeWarning)
+			var data any
+			// Use "blob" as default, as per spec:
+			// https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/binaryType
+			if w.binaryType == "" || w.binaryType == blobBinaryType {
```
by the specification `binaryType` will be `blob` at the beginning. 

Given that we are breaking the behaviour, I am not certain if we want to have at least one k6 version with it being empty and just throwing an exception instead of warning. 

That was kind of the original plan in my mind to let more people know this will be broken, but to be honest I am not certain it would make much of a difference in people updating their code if they don't update k6 all the time. 

So it might be better to just switch over to `blob` by default either way,

joanlopez replied:
Gotcha 👍🏻  I see, so #65 basically

mstoykov commented on websockets/websockets_test.go:
```diff
@@ -315,8 +316,94 @@ func TestBinaryState(t *testing.T) {
+			const sent = new Uint8Array([164,41]).buffer
+			ws.send(sent)
+			ws.onmessage = (e) => {
+				//if (!(e.data instanceof Blob)) {
+				//	throw new Error("Wrong event.data type; expected: Blob, got: "+ typeof e.data)
+				//}
```
I would argue this is blocking - this is *especially* blocking if we want to move this to the core.

JS objects that do not work with instanceof aren't a good thing. K6 in general doesn't care about this as most of the objects we have - are not meant to be constructed from outside, but Blob isn't one of those. Another example of htis will be http.File.

mstoykov replied:
Probably the easiest way to get around this in this case will be https://github.com/dop251/goja/issues/10#issuecomment-274339012 as otherwise the alternative is to create a prototype and reuse it.

Which unfortunately involves a lot more code and all of the examples I can find in goja/sobek are ... way too complicated

joanlopez replied:
I decided to go with what's suggested here: https://github.com/dop251/goja/issues/529

olegbespalov reviewed (commented):
No specific feedback from my side, what Mihails said + linter fixes sounds like to plan to address :+1: 

nit: There are few places where we just throw an error like `common.Throw(rt, err)`, I'd suggest add some context by wrapping the error

joanlopez commented on websockets/websockets_test.go:
```diff
@@ -315,8 +316,94 @@ func TestBinaryState(t *testing.T) {
+func TestBinaryType(t *testing.T) {
+	t.Parallel()
+
+	t.Run("default", func(t *testing.T) {
```
Sure! I don't have strong opinion on this aspect, so I'm happy to follow with whatever is common! :)

olegbespalov commented on websockets/blob.go:
```diff
@@ -0,0 +1,68 @@
+	return b.data.String()
+}
+
+func (r *WebSocketsAPI) blob(call sobek.ConstructorCall) *sobek.Object {
```
Wasn't this supposed to be exported? :thinking:  https://github.com/grafana/xk6-websockets/pull/74#discussion_r1644691748

joanlopez replied:
Ah, I thought that [being part of module `Exports()`](https://github.com/grafana/xk6-websockets/blob/9e992116dc47b0dbb6f3724231dc87bcc689347e/websockets/websockets.go#L48) is enough.
Wdyt? cc/ @mstoykov

olegbespalov replied:
Ah, probably true, it's likely my misunderstanding :+1:

joanlopez replied:
Don't worry, in any case that made me realize that it's probably good idea to also add a couple tests to ensure that the `Blob`'s basic functionality works as expected (https://github.com/grafana/xk6-websockets/pull/74/commits/89d1b52c4359dc55e0d3a6cc295c1256e1ac9b7e) 👌🏻

olegbespalov reviewed (approved):
Looks good to me!

Please also update README.md since it contains some Blob references that become misleading after merging this :+1:

olegbespalov commented on go.mod:
```diff
@@ -1,37 +1,39 @@
+	go.k6.io/k6 v0.51.1-0.20240612130226-3415ce6edc99
 	go.uber.org/goleak v1.3.0
-	gopkg.in/guregu/null.v3 v3.3.0
+	gopkg.in/guregu/null.v3 v3.5.0
```
That could trigger an update in k6 :cry: 

We fixed that in our https://github.com/grafana/k6/blob/master/modtools_frozen.yml#L10-L12 modtools config

Not sure if [we want to update it](https://github.com/grafana/k6/blob/master/Dependencies.md#exceptions), at least probably not from the module :man_shrugging:

joanlopez replied:
Ah right! Yeah, I was struggling while finding _"the right balance of updates"_ between `k6` and `sobek` (because the `latest` of each doesn't work due to some incompatibility with the `ast` package), so I ended up updating these by mistake.

Fixed! Thanks! 🙇🏻

olegbespalov replied:
NP :+1:

olegbespalov reviewed (approved):
:rocket:

mstoykov commented on go.mod:
```diff
@@ -1,37 +1,39 @@
 module github.com/grafana/xk6-websockets
 
-go 1.20
+go 1.21
+
+toolchain go1.22.3
```
Can we revert this as it will propagate through everything? 

I am all for us to go do it in k6 ... soon - probably next cycle, but doing it first in the extensions seems like a bad idea.

mstoykov commented on websockets/blob_test.go:
```diff
@@ -0,0 +1,52 @@
+	ts := newTestState(t)
+	val, err := ts.runtime.RunOnEventLoop(`
+		(async () => {
+          const blob = new Blob(["P", "A", "SS"]);
+		  const reader = blob.stream().getReader();
+		  const {value} = await reader.read();
+		  return value;
```
```suggestion
		  const blob = new Blob(["P", "A", "SS"]);
		  const reader = blob.stream().getReader();
		  const {value} = await reader.read();
		  return value;
```

mstoykov commented on go.mod:
```diff
@@ -42,26 +44,26 @@ require (
+	google.golang.org/genproto/googleapis/api v0.0.0-20240617180043-68d350f18fd4 // indirect
+	google.golang.org/genproto/googleapis/rpc v0.0.0-20240617180043-68d350f18fd4 // indirect
+	google.golang.org/grpc v1.64.0 // indirect
+	google.golang.org/protobuf v1.34.2 // indirect
```
It feels like a bunch of those bumps are because you did `go get -u` instead of because of k6 updates, and I am not certain we need the update of k6 - so maybe just reverting this changes will be better :shrug:

mstoykov commented on websockets/blob.go:
```diff
@@ -0,0 +1,72 @@
+
+	var blobParts []interface{}
+	if len(call.Arguments) > 0 {
+		if parts, ok := call.Arguments[0].Export().([]interface{}); ok {
```
nit: Probably would use `ExportTo` in this case

mstoykov reviewed (changes requested):
LGTM in general but is missing the fairly important part of using Blob to *send* data.

So I will be requesting that as well on top of some nitpicks and the dependency updates :)

joanlopez commented on websockets/blob.go:
```diff
@@ -0,0 +1,178 @@
+}
+
+// toByteSlice converts a slice of numbers to a slice of bytes.
+func toByteSlice(data interface{}) []byte {
```
I'm open for suggestions, although none of the ones I've seen so far really convinces me. If you prefer any other, just tell me.

joanlopez commented:
> LGTM in general but is missing the fairly important part of using Blob to _send_ data.
> 
> So I will be requesting that as well on top of some nitpicks and the dependency updates :)

Hey @mstoykov,

On top of nitpicks, I've added a bunch of changes:
- Extended the support to construct `new Blob`s for all the types described in the [spec](https://developer.mozilla.org/en-US/docs/Web/API/Blob/Blob) (_[ArrayBuffer](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer)s, [TypedArray](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray)s, [DataView](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DataView)s, [Blob](https://developer.mozilla.org/en-US/docs/Web/API/Blob)s, strings, or a mix of any of such elements_)
- Added support for the `type` attribute (now assigned from constructor `options`)
- Added support for the missing methods: `bytes()`, `slice()` and `stream()`.
- Changed the return type of some of these methods to `Promise`, as described in the spec: `arrayBuffer()`, `bytes()` and `text()`.
- Added support to also use `Blob` for `send`ing.

So, with all that, I think we're in a quite decent position (after your review), to ship this support, as I think it's quite compliant.
That said, I'd recommend you to take a look at all these changes by following the commit history in order, as it's probably easier to follow, specially considering that you already did a review. Just in case.

Thanks! 🙇🏻

mstoykov commented on websockets/helpers.go:
```diff
@@ -11,3 +15,51 @@ func must(rt *sobek.Runtime, err error) {
+	return o.Prototype().Get("constructor") == rt.GlobalObject().Get("DataView")
+}
+
+func isBlob(o *sobek.Object, rt *sobek.Runtime) bool {
+	return o.Prototype().Get("constructor") == rt.GlobalObject().Get("Blob")
+}
```
I am pretty sure this only works as in your tests you set Blob to be property of the global object. But that isn't how `import { Blob } from ...` works now or works with Native ESM.

I am not strictly against importing websockets setting `Blob` on the global object. Another alternative is to have the Blob "import" be saved on the module instance and use that to chec if it is the same.

Arguably at some future case where Blob is internal it will (like SetTimeout) be available globally.

joanlopez replied:
Nice catch! 🙇🏻  Let's go with:

> Another alternative is to have the Blob "import" be saved on the module instance and use that to check if it is the same.

As it is still experimental, I think it's better to follow the common procedure, and leave `Blob` only available with the import, and wait until we declare it stable. As you said, the same we did for `setTimeout` et al.

Fixed at https://github.com/grafana/xk6-websockets/pull/74/commits/c38822a0d53ba041d8979a93d31e0ad79ee17de1

mstoykov commented on websockets/websockets.go:
```diff
@@ -448,13 +440,25 @@ func (w *webSocket) queueMessage(msg *message) {
+				return errors.New(binarytypeError)
+			case blobBinaryType:
+				var err error
+				data, err = rt.New(rt.Get("Blob"), rt.ToValue([]interface{}{msg.data}))
```
Same problem as with `isBlob`

joanlopez replied:
Fixed at https://github.com/grafana/xk6-websockets/commit/c38822a0d53ba041d8979a93d31e0ad79ee17de1

mstoykov reviewed (changes requested):
LGTM in general, but the isBlob part is actually blocking.

I will be glad to have more tests especially if we can get some suite working, but we can have this after we merge this.

joanlopez commented on websockets/blob.go:
```diff
@@ -0,0 +1,192 @@
+		for n, part := range blobParts {
+			var err error
+			switch v := part.(type) {
+			case []uint8:
+				_, err = b.data.Write(v)
+			case []int8, []int16, []int32, []int64, []uint16, []uint32, []uint64, []float32, []float64:
+				_, err = b.data.Write(toByteSlice(v))
```
A `[]uint8` can be initialized with `Uint8Array` or `Uint8ClampedArray`, and similarly for the rest of the integer and float-based slices, with the other typed arrays: `Int8Array`, `Int16Array`, `Int32Array`, etc. It's not only described in the spec, but I also did some manual tests with Deno and they're also supported there. 

Sounds good?

joanlopez replied:
Not sure if that's what raised your concern, but I just realized that when I added all the test cases for the typed arrays, I involuntarily messed up the one for `ArrayBuffer`, fixed now at https://github.com/grafana/xk6-websockets/pull/74/commits/ad3597ca8e4d76a473a77e79c298d4a35ef3f664.

That's the difference between receiving a `sobek.ArrayBuffer` or a `[]uint8`.

mstoykov replied:
Apperantly I missed this being changed in https://github.com/dop251/goja/commit/235299346f15aac0ab4e280a73bbbf9747b641d8 

 well okay I guess we need this code :shrug:
